File: Aman/DP Recursion and Backtracking/Minimum Cost For Tickets.cpp
class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        int n=days.size();
        unordered_map<int,int>dp;
        int prev7=0,prev30=0;
        for(int i=0;i<=n;i++)
        {   
           while(days[i]-days[prev7]>=7)prev7++;
           while(days[i]-days[prev30]>=30)prev30++;
            
            dp[i]=min({dp[i-1]+costs[0],dp[prev7-1]+costs[1],dp[prev30-1]+costs[2]});
        }
        return dp[n-1];
    }
};

File: Harish Kumar/Dynamic Programming & Recursion/ Minimum Cost For Tickets/ Minimum Cost For Tickets.java
class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        int dp[] = new int[days.length];
        int ans = helper(days, costs, 0,dp);
        return ans;
    }
    
    private static int helper(int days[], int cost[], int i, int dp[])
    {
        if(i >= days.length)
        {
            return 0;
        }
        
        if(dp[i] > 0)
        {
            return dp[i];
        }
        
        int oneDay = cost[0] + helper(days, cost, i+1,dp);
        
        int k = i;
        
        for(; k < days.length;k++)
        {
            if(days[k] >= days[i]+7)
            {
                break;
            }
        }
        
        int sevenDay = cost[1] + helper(days, cost, k,dp);
        
        for(; k < days.length;k++)
        {
            if(days[k] >= days[i]+30)
            {
                break;
            }
        }
        
        int thirtyDay = cost[2] + helper(days, cost, k, dp);
        
        dp[i] = Math.min(oneDay, Math.min(sevenDay, thirtyDay));
        
        return Math.min(oneDay, Math.min(sevenDay, thirtyDay));
    
        
    }
}
